### ISSUE TITLE
Models should declare their own inference input contract

### ISSUE THREAD

**Issue by robertleist** [Iquana-tool/iquana-tool#27]
# Models should declare their own inference input contract

**Repos affected:** `iquana-toolbox`, `ai-service`, `backend`, `frontend-react`
**Type:** design / refactor · **Priority:** not urgent, but blocks every new in-context method
**Status:** written up after the SAM 3 cross-image `top_k` fix; no implementation yet

## Summary

Each task currently has exactly one hardcoded request shape, one hardcoded set of user-facing
knobs, and one hardcoded piece of frontend that renders them. When a model needs something
different from what its task offers, there is nowhere to say so, so the difference gets pushed
into whichever layer noticed it — a magic constant in the model, a renamed field in the API, or
a knob the user is expected to understand but can't.

Proposal: **a model declares, per task, what it consumes** — an input contract carried on
`ModelInfo`, logged into MLflow with the model, served through the model catalog, and rendered
generically by the frontend. Same mechanism `training_parameters` already uses for the training
UI, extended to inference inputs.

## Background: how we got here

`cross-image-suggestion` transfers a concept onto a target image from exemplars found elsewhere
in the dataset. Its only implementation today is the SAM 3 **concat workaround**: SAM 3 is
intra-image, so `ai-service/models/sam3.py` composites the exemplar's whole source image beside
the target on one canvas and uses the exemplar's mask bbox as a positive visual prompt.

The task's request (`CrossImageSuggestionRequest`) carries a list of exemplars, each an
*instance*: `image_url` + one mask. The backend retrieves the top-k exemplars by a retrieval
strategy and sends them. That is the natural unit for the task and for the retrieval layer,
which ranks contours.

It is not the unit SAM 3 pays in. Each exemplar drags its whole source image onto the canvas,
and the processor then resizes that canvas to a fixed input — so `top_k=5` meant five images
beside the target and a target too small to segment, fatal when instances are small.

Three fixes were tried:

1. **Rename the knob to `max_exemplar_images`, default 1.** Rejected: it makes the user reason
   about a workaround's internal cost. Nobody outside this repo knows why 5 is fine for one
   model and catastrophic for another.
2. **Redefine an exemplar as an image carrying all its masks.** Rejected: it promotes one
   model's constraint into the task's contract. The exemplar *is* the instance; a real
   in-context method consumes instances and would want all of them, from as many images as
   possible.
3. **Hard-wire SAM 3 to `request.exemplars[:1]`** (shipped). Correct as far as it goes — a model
   that can't spend what the task offers should refuse it model-side, leaving the task
   model-agnostic — but it is a silent constant, invisible to the user and to the API.

The knob is still there, still says "Top-k", and is still ignored by the only model that
implements the task.

### 1. The task's request shape is one model's shape, generalized

`cross-image-suggestion` reads "target image + instance exemplars" because that is what the
backend could retrieve, not because it is what in-context methods consume:

| method | conditioning it actually consumes | priced per |
| --- | --- | --- |
| SAM 3 (concat) | reference **images**; instances only supply boxes on the pasted tile | image |
| NTT / Matcher / PerSAM / INSID3 | instance **masks**, across as many images as possible | instance |
| prototype / embedding methods | **vectors** — the pgvector store already has them; shipping pixels is pure waste | vector |
| text-promptable | a **concept name**, zero exemplars | — |
| finetuned specialist | nothing but the target image | — |

One request type cannot be honest about all five. Today every model un-mangles the one shape it
is given.

### 2. The unit of "how many" is model-specific, but the knob is user-facing

`top_k` means "top-k exemplar instances" to the task, "top-k images" to SAM 3, and nothing at
all to a text-promptable model. It is rendered as one number input for every cross-image model
(`frontend-react/src/components/inference/LabelModelPlanner.jsx`), so it is wrong for at least
two of the three.

### 3. Partial exemplar coverage is a correctness problem the contract can't express

For an image-conditioned model, pasting a reference image in which only *some* instances of the
concept are marked is worse than not pasting it: the model sees unprompted objects that look
exactly like the concept and learns the wrong boundary. Retrieval returns ranked instances with
no guarantee that an image's other instances came along.

The information exists — `Masks.fully_annotated`, review status — but there is no way for a
model to say "I need reference images whose annotation of this concept is complete", so
retrieval cannot rank by it.

### 4. Per-task fields are hardcoded through every layer

Adding one input to one task today means editing, in order:

- `iquana-toolbox` — the task's request schema
- `ai-service` — the handler
- `backend/app/schemas/inference.py` — `InferenceStepRequest` gains a field, plus a validator
  (`_require_strategy_for_cross_image`) asserting task-specific presence rules
- `backend/app/schemas/cross_image.py` — the interactive route's body
- `backend/app/services/cross_image_orchestration.py` — request assembly
- `frontend-react` — `isCrossImage = selected?.task === "cross-image-suggestion"` and a
  hand-written input per field

Five repos-worth of edits to expose one number, and the result still can't vary per model.

### 5. Tunables are stringly-typed tags

`threshold`, `mask_threshold`, `min_target_frac` are documented in `usage_tip` prose and
`ModelInfo.tags` (`"threshold": "0.3"`), passed as an untyped `params` dict, and exposed to the
user nowhere (see also #13). There is no defaults/range/type information to render a control from.

## Proposed solution: model-declared input contracts, generically rendered

Each model declares, per task it advertises, an **input contract**: the conditioning it
consumes, the cardinality it accepts, and the tunables it exposes. The frontend fetches it and
renders the inputs; the backend reads it to decide what to retrieve and validates against it;
the model receives only what it asked for.

### This pattern already exists here

`ModelInfo.training_parameters: List[HyperParameter]` — *"The frontend renders the training
config UI generically from these (defaults, ranges, options)."* `HyperParameter` carries `key`,
`label`, `type`, `default_value`, `options`, `min_value`, `max_value`, `step`, `description`:
enough to render a control without the frontend knowing what the parameter means.

The proposal is that same idea applied to inference inputs, with one addition training doesn't
need: a declaration of what *conditioning data* the model consumes, so the backend knows what to
retrieve and resolve.

`app/services/exemplar_retrieval.py` is the second precedent: strategies declare
`required_kinds`, and `strategy_options()` marks a strategy unavailable — with a reason — on a
dataset that lacks those embeddings. Declared requirements, checked against reality, surfaced to
the UI. Input contracts should compose with it: a contract requiring `region_mean` vectors
narrows the usable strategies the same way.

### Shape sketch

```python
class ConditioningSpec(BaseModel):
    """What conditioning data a model consumes for one task."""
    kind: Literal["instances", "reference_images", "embeddings", "concept_text", "none"]
    unit: Literal["instance", "image", "vector"]      # what max_units counts
    min_units: int = 0
    max_units: int | None = None                      # None = unbounded; SAM 3 concat: 1
    requires_complete_annotation: bool = False        # reference images must be fully annotated
    embedding_kinds: list[str] = []                   # e.g. ["region_mean"], ties into retrieval
    user_selectable_count: bool = True                # False => don't render a count knob at all


class InputContract(BaseModel):
    task: str
    conditioning: ConditioningSpec
    parameters: list[HyperParameter] = []             # threshold, mask_threshold, min_target_frac
    notes: str | None = None                          # rendered as help text, not as an excuse
```

`ModelInfo` gains `input_contracts: list[InputContract]`, one per advertised task. It is already
`model_dump()`-ed into MLflow metadata at registration, so it travels with the model.

### Worked examples

**SAM 3, cross-image** — the constraint stops being a silent slice and becomes a fact the whole
stack can see:

```python
InputContract(
    task="cross-image-suggestion",
    conditioning=ConditioningSpec(
        kind="reference_images", unit="image", min_units=1, max_units=1,
        requires_complete_annotation=True, user_selectable_count=False,
    ),
    parameters=[
        HyperParameter(key="threshold", label="Detection sensitivity", type="float",
                       default_value=0.3, min_value=0.0, max_value=1.0, step=0.05),
        HyperParameter(key="min_target_frac", label="Must sit on target", type="float",
                       default_value=0.5, min_value=0.0, max_value=1.0, step=0.05),
    ],
    notes="Concat workaround: the reference image is pasted beside the target, so only one "
          "fits before the target loses resolution.",
)
```

The count knob disappears from the UI instead of lying; the backend retrieves one image and
skips resolving four exemplars nobody will use; retrieval can prefer fully-annotated images.

**An instance-conditioned method** — `kind="instances"`, `unit="instance"`, `max_units=None`,
`user_selectable_count=True`. The same "Top-k" knob appears, and now means what it says.

**A text-promptable model** — `kind="concept_text"`. No exemplar retrieval runs at all, no
strategy picker renders, and the model stops being offered on datasets with no annotations for
reasons that have nothing to do with it.

### What this deletes

- `request.exemplars[:1]` in `ai-service/models/sam3.py`
- `isCrossImage`-style task branching in `LabelModelPlanner.jsx`
- `retrieval_strategy` / `top_k` as hardcoded fields on `InferenceStepRequest`, and
  `_require_strategy_for_cross_image`
- the `usage_tip` sentences that exist only to explain a knob's real meaning

## Implementation sketch

1. **Toolbox** — `ConditioningSpec`, `InputContract`, `ModelInfo.input_contracts`. Additive and
   defaulted, so nothing breaks before models populate it.
2. **ai-service** — each model declares contracts for the tasks it mixes in; `register_task`
   could carry a *default* contract per task so an undeclared model still resolves to today's
   behaviour. Handlers assert against their own contract instead of defensively slicing.
3. **backend** — model catalog serves contracts alongside `registry_key` / `task` / `label_ids`;
   step schemas carry a generic `inputs: dict[str, Any]` validated against the contract rather
   than a fixed field list; `cross_image_orchestration` reads `conditioning` to decide what to
   retrieve, how much, and in which unit.
4. **frontend** — one generic renderer, driven by the contract, replacing the per-task blocks;
   reuse whatever renders `training_parameters`.
5. **Fallback + rollout** — models already registered in MLflow have no contract in their
   metadata, so per-task defaults must reproduce current behaviour exactly. Note the
   registration gotcha: `MLFlowModelRegistry.register_model` **skips re-logging an
   already-registered base model**, so contract changes need the registered model deleted to
   force a clean re-log; restarting the service is not enough.

## Cost and risks

- Genuinely cross-repo: toolbox schema, every model's declaration, backend catalog + validation,
  a new frontend renderer. This is the expensive option — it is proposed because the cheap ones
  were tried above and each one lied to someone.
- **Contract drift**: a declaration that disagrees with the handler is worse than none. The
  handler should validate against its own contract so drift fails loudly on the first request.
- **Versioning**: contracts live in MLflow metadata per model *version*; a model re-logged with
  a different contract while a batch job is mid-run must not change that job's meaning.
- **Over-generalisation**: `kind` must stay a small closed enum. The moment it becomes a
  free-form schema language, every consumer has to interpret it and the win evaporates.

## Open questions

- Contract per (model, task) or per task with per-model overrides? SAM 3 advertises three tasks
  and only one of them has an unusual contract.
- Does `params` stay a free dict at the wire level, with the contract only describing how to
  render it, or does the contract become the validation authority?
- Who owns "how many" when both sides have an opinion — a model that accepts up to 8 instances
  and a user who wants 20?
- Should retrieval strategies advertise which `kind`s they can *produce*, so strategy
  availability follows from the contract instead of being filtered separately?
- Is `requires_complete_annotation` a hard filter (only fully-annotated images are eligible) or a
  ranking preference with a warning?

## Acceptance criteria

- [ ] A new in-context method with a different conditioning kind can be added by writing the
      model and its contract — no backend field, no frontend component, no task branch.
- [ ] No user-facing control whose meaning depends on which model is selected.
- [ ] SAM 3's one-image limit is expressed as data, and the backend stops retrieving and
      resolving exemplars that will be discarded.
- [ ] Models with no declared contract behave exactly as they do today.
- [ ] `threshold` / `mask_threshold` / `min_target_frac` are reachable from the UI with types,
      defaults and ranges.

## Out of scope

- Splitting `cross-image-suggestion` into instance-conditioned and image-conditioned tasks. Worth
  revisiting once a real instance-conditioned method exists and the request shapes genuinely
  diverge; a contract may make the split unnecessary.
- Changing the retrieval strategies themselves.
- User-uploaded model weights (separate model-zoo work).

## Anchors

- `ai-service/models/sam3.py` — `suggest_cross_image`, the hard-wired `exemplars[:1]`
- `ai-service/models/base.py` — `register_task(name, RequestType, handler)`, capability mixins
- `iquana-toolbox/src/iquana_toolbox/schemas/model_info.py` — `ModelInfo`, `training_parameters`
- `iquana-toolbox/src/iquana_toolbox/schemas/training.py` — `HyperParameter`, the precedent
- `iquana-toolbox/src/iquana_toolbox/schemas/networking/http/services.py` — `CrossImageExemplar`,
  `CrossImageSuggestionRequest`
- `backend/app/services/exemplar_retrieval.py` — strategy registry, `required_kinds`
- `backend/app/services/cross_image_orchestration.py` — request assembly
- `backend/app/schemas/inference.py` — `InferenceStepRequest`, the per-task fields
- `frontend-react/src/components/inference/LabelModelPlanner.jsx` — the hardcoded task branch
